docs/architecture/adr-059-go-api-stability.md

```diff
@@ -0,0 +1,197 @@
+
+* **External package:** a different Go package, can be a child or sibling package in the same project.
+
+* **Internal code:** code not intended for use in external projects.
+
+* **Internal directory:** code under `internal/` which cannot be imported in external projects.
+
+* **Exported:** a Go identifier starting with an uppercase letter, which can therefore be accessed by an external package.
+
+* **Private:** a Go identifier starting with a lowercase letter, which therefore cannot be accessed by an external package unless via an exported field, variable, or function/method return value.
+
+* **Public API:** any Go identifier that can be imported or accessed by an external project, except test code in `_test.go` files.
+
+* **Private API:** any Go identifier that is not accessible via a public API, including all code in the internal directory.
+
+## Alternative Approaches
+
+- Split all public APIs out to separate Go modules in separate Git repositories, and consider all Tendermint code internal and not subject to API backwards compatibility at all. This was rejected, since it has been attempted by the Tendermint project earlier, resulting in too much dependency management overhead.
+
+- Simply document which APIs are public and which are private. This is the current approach, but users should not be expected to self-enforce this, the documentation is not always up-to-date, and external projects will often end up depending on internal code anyway.
+
+## Decision
+
+From Tendermint 1.0, all internal code (except private APIs) will be placed in a root-level [`internal` directory](https://golang.org/cmd/go/#hdr-Internal_Directories), which the Go compiler will block for use by external projects. All exported items outside of the `internal` directory are considered a public API and subject to backwards compatibility guarantees, except files ending in `_test.go`.
+
+The `crypto` package will be split out to a separate module in a separate repo. This is the main general-purpose package used by external projects, and is the only Tendermint dependency in e.g. IAVL which can cause some problems for projects depending on both IAVL and Tendermint.
+
+The `tm-db` package will remain a separate module in a separate repo.
+
+## Detailed Design
+
+### Public API
+
+When preparing our public API for 1.0, we should keep these principles in mind:
+
+- Limit the number of public APIs that we start out with - we can always add new APIs later, but we can't change or remove APIs once they're made public.
+
+- Before an API is made public, do a thorough review of the API to make sure it covers any future needs, can accomodate expected changes, and follows good API design practices.
+
+The following current packages will be part of our public API in 1.0 in some form:
+
+- `abci`
+- `config`
+- `crypto` (separate repository)
+- `libs/log`
+- `light`
+- `node`
+- `privval` (client parts)
+- `rpc/client`
+- `types` (respective packages)
```

Comment:
have you thought about how we're going to split types and move them to respective packages? or this won't happen before 2.0?